### ISSUE TITLE
Import static with protected base class yiels wrong class reference compiled in resulting classfile

### ISSUE THREAD

**Issue by laeubi** [eclipse-jdt/eclipse.jdt.core#809]
I have a quite uncommon class structure (but I can't change it as the code is generated), where I have a **package protected class** that has a **public static method**, then there is another **public class** that extends the first one:

![grafik](https://user-images.githubusercontent.com/1331477/222908414-99809443-53f7-424f-aa66-3c1b72886852.png)

Now I have a class in another package that uses this public class and calls the static method and has a static import with wildcard:

![grafik](https://user-images.githubusercontent.com/1331477/222908465-fe7259fe-2da8-4b1d-aa22-6bbde6b0f07d.png)

in the IDE this compiles fine, no warnings or errors:

![grafik](https://user-images.githubusercontent.com/1331477/222908508-382168b0-2b22-493f-ba04-b481f151ec6b.png)

But if I now run the class it fails:

> Exception in thread "main" java.lang.IllegalAccessError: failed to access class test.access.staticprotected.ThisIsProtected from class test.another.pkg.TestMe (test.access.staticprotected.ThisIsProtected and test.another.pkg.TestMe are in module test.access.staticprotected of loader 'app')

The important part here is that **test.access.staticprotected.ThisIsProtected** is used, what is forbidden!

Interesting enough, this can be also seen in the IDE if i organize the imports what results in:
![grafik](https://user-images.githubusercontent.com/1331477/222908697-10ff9efd-5bef-4f21-89d4-d2f6735644c7.png)

If I now change the code to use the public class like this
![grafik](https://user-images.githubusercontent.com/1331477/222908974-3c520b83-9727-4a93-9405-0122b292cc19.png)

Then the IDE is fine **but** running it results in the same runtime error.

Next I change it to use full qualified name like this
![grafik](https://user-images.githubusercontent.com/1331477/222909697-5175c9e2-7419-403f-8bc6-357156afaf8a.png)

Now it works in the IDE **and** if I run the program.

The sourceode for this example can be found here: [test.access.staticprotected.zip](https://github.com/eclipse-jdt/eclipse.jdt.core/files/10888888/test.access.staticprotected.zip)

@stephan-herrmann I think we can need your compiler expertise here, I can't even imagine if this is a bug or probably expected what one is seeing here, but whatever this is there seems a discrepancy between what the IDE thinks about this code and what the runtime is assuming.

**Comment by HannesWell** [comment]
@mpalat can you maybe have a look at this as well? :)

**Comment by stephan-herrmann** [comment]
Let's see if an answer to this issue exists.

### First check: what happens when using *javac* instead of *ecj*:

ecj generates: `invokestatic test/access/staticprotected/ThisIsProtected.thisIsPublic:()Ljava/lang/String;`
This throws `IllegalAccessError`.

javac generates: `invokestatic test/access/staticprotected/ThisIsPublic.thisIsPublic:()Ljava/lang/String;`
This doesn't throw.

### Is throwing `IllegalAccessError` justified?

[JVMS 5.4.4](https://docs.oracle.com/javase/specs/jvms/se19/html/jvms-5.html#jvms-5.4.4) handles access to classes separately from access to methods. If `invokestatic` is considered only as accessing a method, then the fact that the method is `public` satisfies the access check. 

Do we also need to check access to the class used? Obviously, `ThisIsPublic` is accessible, whereas `ThisIsProtected` is not. But is this relevant for `invokestatic`?

Is the jvm at fault? For comparison I ran the ecj-generated classfiles with a J9 VM, which raised the same `IllegalAccessError`, this is the VM I happened to find on my disk:

```
openjdk version "10.0.2-adoptopenjdk" 2018-07-17
OpenJDK Runtime Environment (build 10.0.2-adoptopenjdk+13)
Eclipse OpenJ9 VM (build openj9-0.9.0, JRE 10 Linux amd64-64-Bit Compressed References 20180813_102 (JIT enabled, AOT enabled)
OpenJ9   - 24e53631
OMR      - fad6bf6e
JCL      - 7db90eda56 based on jdk-10.0.2+13)
```
Looking deeper into JVMS revealed [JVMS 5.4.3.3](https://docs.oracle.com/javase/specs/jvms/se19/html/jvms-5.html#jvms-5.4.3.3):

> To resolve an unresolved symbolic reference from D to a method in a class C, the symbolic reference to C given by the method reference is first resolved ([§5.4.3.1](https://docs.oracle.com/javase/specs/jvms/se19/html/jvms-5.html#jvms-5.4.3.1)). Therefore, any exception that can be thrown as a result of failure of resolution of a class reference can be thrown as a result of failure of method resolution.


This justifies the error thrown when resolving the Methodref as created by ecj.

### Which bytecode *should* compilers generate?

javac generates this constant pool entry:
```
   #2 = Methodref          #14.#15        // test/access/staticprotected/ThisIsPublic.thisIsPublic:()Ljava/lang/String;
```

About this kind of entry [JVMS 4.4.2](https://docs.oracle.com/javase/specs/jvms/se19/html/jvms-4.html#jvms-4.4.2) states:
```
CONSTANT_Methodref_info {
    u1 tag;
    u2 class_index;
    u2 name_and_type_index;
}
```
and then:

>class_index
>
>   The value of the class_index item must be a valid index into the constant_pool table. The constant_pool entry at that index must be a CONSTANT_Class_info structure (§4.4.1) representing a class or interface type that has the field or method as a member.


**Question**: Does class `ThisIsPublic` have method `thisIsPublic()` _as a member_?

Over to JLS: 

[JLS 8.2](https://docs.oracle.com/javase/specs/jls/se19/html/jls-8.html#jls-8.2):
> The members of a class are all of the following:
>
>    Members inherited from its direct superclass type (§8.1.4), ...


**Question**: Does  `ThisIsPublic` *inherit* method `thisIsPublic()`?

It's always good to check *which concept* grants visibility to a given member, so we look at 
[JLS 8.4.8](https://docs.oracle.com/javase/specs/jls/se19/html/jls-8.html#jls-8.4.8) and indeed:
> A class C inherits from its direct superclass type D all concrete methods m (both static and instance) for which all of the following are true:  ...

As all criteria mentioned are fulfilled, yes, `ThisIsPublic` *inherits* method `thisIsPublic()`. So the bytecode generated by javac seems to be legal.

**Question**: Is the legal bytecode mandated?

I could not find any binding answer neither in JLS nor JVMS, but I might be missing something. Still I'd say the above analysis *suggests* that generating a Methodref via `ThisIsPublic` is *nicer* than what ecj does.

### What about "The static member type {0}.{1} should be accessed directly"?

Ecj raises this warning since version 3.0. Unfortunately the commit introducing this warning is not linked to any bug or other discussion that would reveal the background. So all we have is the javadoc of `org.eclipse.jdt.core.JavaCore.COMPILER_PB_INDIRECT_STATIC_ACCESS`:
```
 Compiler option ID: Reporting Indirect Reference to a Static Member.
 When enabled, the compiler will issue an error or a warning whenever a static field
    or method is accessed in an indirect way. A reference to a static member should
    preferably be qualified with its declaring type name.
```

* So, what's the origin of this "should preferably"?
* Was it relevant at that time?
* Is it still relevant?
* Was it a misunderstanding from the beginning?

### TL;DR:

* Throwing `IllegalAccessError` as a result of the bytecode generated by ecj seems correct.
* I could not find a specification rule defining ecj's behavior as incorrect.
* Changing the behavior should still be considered as a service to users.
* It's unclear, why ecj (historically) prefers (in code generation and via warning to users) access to static members via the declaring class.

**Comment by stephan-herrmann** [comment]
PS: Yes, that code structure is "crude"  :smile:

**Comment by laeubi** [comment]
@stephan-herrmann I thing giving preference to the "lower type" and a warning would be fine *if* the method would be accessible that way but as you said it is not! So i think this is "illegal" implicitly without stating it explicit because it will result in an access error!

So lets say I try to call a _private_ method then for sure I exspect the compiler to complain and not to generate a bytecode that will give illegalaccess error at runtime!

The warning has another background (and as said, one don't get any warnings here!) because I can do the follwoing:


```
public A extends B {

}

public B {
  public static String howdy() {
    ...
  }
}
```

Now I call it like this `A.howdy();` EJC will generate a warning and thats actually good, because if the above code evolves, something like this can happen:

```
public A extends B {
  public static int howdy() {
    ...
  }
}

public B {
  public static String howdy() {
    ...
  }
}
```

So now my compiled binary references a method with a different signature and will fail with (I think) incompatible class change error!


But here we have the situation of:
```
public A extends B {

}

B {
  public static String howdy() {
    ...
  }
}
```

so  `A.howdy();` is the only possible way to access that method and I should *not* get a warning ....


> code structure is "crude" smile

The code is generated by `jextract` (part of java 19) so expect other to complain about this behavior soon, I just _extracted_ this sample from the more complex code to provide you with a simple reproducer of the Probelm so this hopefully can be fixed before it becomes a major complain.

For the background why `jextract` is doing that:

- it takes a header file and generate a class from it (e.g. `mylib.h` --> `mylib_h.java`)
- this can result in very very very large classes that hit the constant-pool limit
- beside that loading such a class creates an enormous amount of method-handles and takes very long even though only a few methods are usually used directly 
- therefore jextract split the class into several `mylib_h.java`, `mylib_h_1.java`, `mylib_h_2.java`, ... where only `mylib_h.java` is public, the rest is package private and `mylib_h` extends `mylib_h_1` extends `mylib_h_2` extends .... and each contains a small set of constants and/or static functions

**Comment by HannesWell** [comment]
Thank you for your analysis :)

> PS: Yes, that code structure is "crude" 😄

It is. If you want a little bit more context, it is generated by `jextract` for the `opengl` example
https://github.com/openjdk/jextract/tree/master/samples/opengl
from the [freeglut.h](https://github.com/FreeGLUTProject/freeglut/blob/master/include/GL/freeglut.h)

**Comment by laeubi** [comment]
There is now a full public example available here that shows the complex case:
- https://github.com/eclipse-platform/eclipse.platform.swt/pull/571

**Comment by laeubi** [comment]
I have created two other issues to address the false warning and the wrong organized imports:

- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/813
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/814

It would be great if this can be fixed as this is currently blocking some new features in the current release.

**Comment by stephan-herrmann** [comment]
Actually, the code already makes efforts towards avoiding the use of inaccessible classes. See `CodeStream.getConstantPoolDeclaringClass(..)` (both overloads).

**What we know**:
* the JVM is correct in throwing `IllegalAccessError`
* ecj makes efforts to use the method's `declaringClass` rather than the class mentioned in source code (supposed to be known as `actualReceiverType`)
* ecj makes additional efforts to switch back to the `actualReceiverType` if needed

**What we don't know**:
* I haven't found any specification positively defining which class to use when generating `invokestatic`.
* I haven't found the rationale, *why* ecj generally prefers using the declaring class
* I haven't found why the code in `getConstantPoolDeclaringClass()` doesn't work as expected in this issue
* I have no clue about why `getConstantPoolDeclaringClass()`  checks compliance for `JDK1_2` and `JDK1_4` to decide the strategy.
  * Did we discuss dropping support for some ancient Java versions? What's the status there?

I'm afraid if we can't answer a majority of the above unknowns, we're likely to fix one problem and create several new problems.

**Comment by stephan-herrmann** [comment]
This burst of commits from 2008 seems to be the start of tweaking class references in field/method access:
* d35754bf597e1e3c63426404dbcb66843b76444c
* 889cbee9e452a2b606e084167406ae9a949acd46
* 81c2952f612a9031156c7da5b3efa0e0eeba128c
* 3de30b43323a1cec0084dbcd8ad69f03637e9a4f

These link to [bug 128563](https://bugs.eclipse.org/bugs/show_bug.cgi?id=128563), [bug 247292](https://bugs.eclipse.org/bugs/show_bug.cgi?id=247292) and [bug 247307](https://bugs.eclipse.org/bugs/show_bug.cgi?id=247307), where more background can be found.

Interestingly, this started out without any consideration about static.

**Comment by iloveeclipse** [comment]
> * Did we discuss dropping support for some ancient Java versions? What's the status there?

See https://github.com/eclipse-jdt/eclipse.jdt.core/discussions/820

**Comment by jzy3d** [comment]
I'd like to confirm that this issue is annoying when using Eclipse with code generated by JExtract. I am quite sure that this new tool developped at Oracle will gain in popularity and think that would be relevant to have Eclipse able to support this.

**Comment by laeubi** [comment]
@stephan-herrmann this is the issue we talked this morning at [EclipseCon2023](https://www.eclipsecon.org/2023), other related ones are:

- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/813
- https://github.com/eclipse-jdt/eclipse.jdt.core/issues/814

[JExtract](https://github.com/openjdk/jextract) is a really cool tool and now that Java 21 is GA with quite stable API in that regard will be a good alternative to writing JNI wrappers, so it would be great if one can use JDT to work with such code as well and is not enforced to use javac....
